server/bleep/src/semantic/chunk.rs

```diff
@@ -76,6 +85,62 @@ pub fn tree_sitter<'a, 'l>(src: &'a str, lang_id: &'l str) -> Result<Vec<Chunk<'
     Ok(chunks)
 }
 
+fn point(src: &str, byte: usize) -> Point {
+    let line = src.as_bytes()[..byte].iter().filter(|&&b| b == b'\n').count();
+    let column = if let Some(last_nl) = src[..byte].rfind('\n') {
+        byte - last_nl
+    } else {
+        byte
+    };
+    Point { byte, column, line }
+}
+
+fn add_chunks<'s>(results: &mut Vec<Chunk<'s>>, src: &'s str, offsets: &[(usize, usize)], start: usize, end: usize, _max_tokens: usize) {
+    //TODO: split if more than _max_tokens
+    let startpoint = point(src, offsets[start].0);
+    let endpoint = point(src, offsets[end].1);
+    if endpoint.byte > startpoint.byte { 
+        results.push(Chunk::new(&src[startpoint.byte .. endpoint.byte], startpoint, endpoint));        
+    }
+}
+
+/// This tries to split the code by lines and add as much tokens as possible until reaching
+/// `max_tokens`. Then it'll reduce to the last newline.
+pub fn by_tokens<'s>(
+    src: &'s str,
+    tokenizer: &Tokenizer,
+    max_tokens: usize,
+    max_lines: usize,
+) -> Vec<Chunk<'s>> {
+    if src.is_empty() {
+        return Vec::new();
+    }
+    let Ok(encoding) = tokenizer.encode(src, false)
+    else {
+        warn!("Could not encode \"{}\"", src);
+        return trivial(src, max_lines);
+    };
+    // FIXME: This unfortunately is the char offset, not the byte offset!
+    // We need to take care of non-ascii unicode at some point
+    // another reason for aleph-alpha-tokenizer...
+    let offsets = encoding.get_offsets();
+    let offset_len = offsets.len();
+    let mut first_newline = 0;
+    let mut last_newline = 0;
+    let mut result = Vec::new();
+    for newline in (0..(offset_len - 1)).filter(|&i| src[offsets[i].1 .. offsets[i + 1].0].contains('\n')) {
+        if newline - first_newline > max_tokens {
+            add_chunks(&mut result, src, offsets, first_newline, last_newline, max_tokens);
+            first_newline = (last_newline + 1).min(offset_len);
+        }
+        last_newline = newline;
+    }
```

Comment:
could you help me understand what is going on here? specifically, how does this check: `newline - first_newline > max_tokens` work?

Reply:
Sure! So our chunks have a size limit: max_tokens. We want to fit as many lines as we can into the chunk. So we have to catch the first occasion where we have more than `max_tokens` tokens, go back one line (that's why we keep the `last_newline` around) and chunk until then. If we added a chunk we reset our `first_newline` to the position after the lines we just put into the chunk.

Reply:
understood, makes sense now.